File: objectModel/Python/cdm/persistence/cdmfolder/attribute_context_persistence.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Optional

from cdm.enums import CdmObjectType, CdmAttributeContextType
from cdm.objectmodel import CdmCorpusContext, CdmAttributeContext
from cdm.persistence import PersistenceLayer
from cdm.utilities import CopyOptions, ResolveOptions, copy_data_utils

from . import utils
from .attribute_context_reference_persistence import AttributeContextReferencePersistence
from .attribute_group_reference_persistence import AttributeGroupReferencePersistence
from .attribute_reference_persistence import AttributeReferencePersistence
from .entity_reference_persistence import EntityReferencePersistence
from .types import AttributeContext


map_type_name_to_enum = {
    'entity': CdmAttributeContextType.ENTITY,
    'entityReferenceExtends': CdmAttributeContextType.ENTITY_REFERENCE_EXTENDS,
    'attributeGroup': CdmAttributeContextType.ATTRIBUTE_GROUP,
    'attributeDefinition': CdmAttributeContextType.ATTRIBUTE_DEFINITION,
    'addedAttributeSupporting': CdmAttributeContextType.ADDED_ATTRIBUTE_SUPPORTING,
    'addedAttributeIdentity': CdmAttributeContextType.ADDED_ATTRIBUTE_IDENTITY,
    'addedAttributeExpansionTotal': CdmAttributeContextType.ADDED_ATTRIBUTE_EXPANSION_TOTAL,
    'addedAttributeSelectedType': CdmAttributeContextType.ADDED_ATTRIBUTE_SELECTED_TYPE,
    'generatedRound': CdmAttributeContextType.GENERATED_ROUND,
    'generatedSet': CdmAttributeContextType.GENERATED_SET,
    'projection': CdmAttributeContextType.PROJECTION,
    'source': CdmAttributeContextType.SOURCE,
    'operations': CdmAttributeContextType.OPERATIONS,
    'operationAddCountAttribute': CdmAttributeContextType.OPERATION_ADD_COUNT_ATTRIBUTE,
    'operationAddSupportingAttribute': CdmAttributeContextType.OPERATION_ADD_SUPPORTING_ATTRIBUTE,
    'operationAddTypeAttribute': CdmAttributeContextType.OPERATION_ADD_TYPE_ATTRIBUTE,
    'operationExcludeAttributes': CdmAttributeContextType.OPERATION_EXCLUDE_ATTRIBUTES,
    'operationArrayExpansion': CdmAttributeContextType.OPERATION_ARRAY_EXPANSION,
    'operationCombineAttributes': CdmAttributeContextType.OPERATION_COMBINE_ATTRIBUTES,
    'operationRenameAttributes': CdmAttributeContextType.OPERATION_RENAME_ATTRIBUTES,
    'operationReplaceAsForeignKey': CdmAttributeContextType.OPERATION_REPLACE_AS_FOREIGN_KEY,
    'operationIncludeAttributes': CdmAttributeContextType.OPERATION_INCLUDE_ATTRIBUTES,
    'operationAddAttributeGroup': CdmAttributeContextType.OPERATION_ADD_ATTRIBUTE_GROUP,
    'unknown': CdmAttributeContextType.UNKNOWN
}

map_enum_to_type_name = {
    CdmAttributeContextType.ENTITY: 'entity',
    CdmAttributeContextType.ENTITY_REFERENCE_EXTENDS: 'entityReferenceExtends',
    CdmAttributeContextType.ATTRIBUTE_GROUP: 'attributeGroup',
    CdmAttributeContextType.ATTRIBUTE_DEFINITION: 'attributeDefinition',
    CdmAttributeContextType.ADDED_ATTRIBUTE_SUPPORTING: 'addedAttributeSupporting',
    CdmAttributeContextType.ADDED_ATTRIBUTE_IDENTITY: 'addedAttributeIdentity',
    CdmAttributeContextType.ADDED_ATTRIBUTE_EXPANSION_TOTAL: 'addedAttributeExpansionTotal',
    CdmAttributeContextType.ADDED_ATTRIBUTE_SELECTED_TYPE: 'addedAttributeSelectedType',
    CdmAttributeContextType.GENERATED_ROUND: 'generatedRound',
    CdmAttributeContextType.GENERATED_SET: 'generatedSet',
    CdmAttributeContextType.PROJECTION: 'projection',
    CdmAttributeContextType.SOURCE: 'source',
    CdmAttributeContextType.OPERATIONS: 'operations',
    CdmAttributeContextType.OPERATION_ADD_COUNT_ATTRIBUTE: 'operationAddCountAttribute',
    CdmAttributeContextType.OPERATION_ADD_SUPPORTING_ATTRIBUTE: 'operationAddSupportingAttribute',
    CdmAttributeContextType.OPERATION_ADD_TYPE_ATTRIBUTE: 'operationAddTypeAttribute',
    CdmAttributeContextType.OPERATION_EXCLUDE_ATTRIBUTES: 'operationExcludeAttributes',
    CdmAttributeContextType.OPERATION_ARRAY_EXPANSION: 'operationArrayExpansion',
    CdmAttributeContextType.OPERATION_COMBINE_ATTRIBUTES: 'operationCombineAttributes',
    CdmAttributeContextType.OPERATION_RENAME_ATTRIBUTES: 'operationRenameAttributes',
    CdmAttributeContextType.OPERATION_REPLACE_AS_FOREIGN_KEY: 'operationReplaceAsForeignKey',
    CdmAttributeContextType.OPERATION_INCLUDE_ATTRIBUTES: 'operationIncludeAttributes',
    CdmAttributeContextType.OPERATION_ADD_ATTRIBUTE_GROUP: 'operationAddAttributeGroup',
    CdmAttributeContextType.UNKNOWN: 'unknown'
}


class AttributeContextPersistence:
    @staticmethod
    def from_data(ctx: CdmCorpusContext, data: AttributeContext) -> Optional[CdmAttributeContext]:
        if data is None:
            return None

        attribute_context = ctx.corpus.make_object(CdmObjectType.ATTRIBUTE_CONTEXT_DEF, data.name)
        attribute_context.type = map_type_name_to_enum[data.type]

        if data.get('parent'):
            attribute_context.parent = AttributeContextReferencePersistence.from_data(ctx, data.get('parent'))

        if data.get('explanation'):
            attribute_context.explanation = data.get('explanation')

        if data.get('definition'):
            if attribute_context.type == CdmAttributeContextType.ENTITY or attribute_context.type == CdmAttributeContextType.ENTITY_REFERENCE_EXTENDS:
                attribute_context.definition = EntityReferencePersistence.from_data(ctx, data.definition)
            elif attribute_context.type == CdmAttributeContextType.ATTRIBUTE_GROUP:
                attribute_context.definition = AttributeGroupReferencePersistence.from_data(ctx, data.definition)
            elif attribute_context.type == CdmAttributeContextType.ADDED_ATTRIBUTE_SUPPORTING \
                    or attribute_context.type == CdmAttributeContextType.ADDED_ATTRIBUTE_IDENTITY \
                    or attribute_context.type == CdmAttributeContextType.ADDED_ATTRIBUTE_EXPANSION_TOTAL \
                    or attribute_context.type == CdmAttributeContextType.ADDED_ATTRIBUTE_SELECTED_TYPE \
                    or attribute_context.type == CdmAttributeContextType.ATTRIBUTE_DEFINITION:
                attribute_context.definition = AttributeReferencePersistence.from_data(ctx, data.definition)

        # I know the trait collection names look wrong. but I wanted to use the def baseclass
        applied_traits = utils.create_trait_reference_array(ctx, data.get('appliedTraits'))
        attribute_context.exhibits_traits.extend(applied_traits)

        if data.get('contents'):
            if attribute_context.contents is None:
                attribute_context.contents = []

            for elem in data.contents:
                if isinstance(elem, str):
                    attribute_context.contents.append(AttributeReferencePersistence.from_data(ctx, elem))
                else:
                    attribute_context.contents.append(AttributeContextPersistence.from_data(ctx, elem))

        return attribute_context

    @staticmethod
    def to_data(instance: CdmAttributeContext, res_opt: ResolveOptions, options: CopyOptions) -> AttributeContext:
        result = AttributeContext()

        exhibits_traits = [trait for trait in instance.exhibits_traits if not trait.is_from_property]

        result.explanation = instance.explanation
        result.name = instance.name
        result.type = map_enum_to_type_name[instance.type]
        result.parent = AttributeContextReferencePersistence.to_data(instance.parent, res_opt, options) if instance.parent is not None else None
        result.definition = PersistenceLayer.to_data(instance.definition, res_opt, options, PersistenceLayer.CDM_FOLDER) if instance.definition is not None else None
        # I know the trait collection names look wrong. but I wanted to use the def baseclass
        result.appliedTraits = copy_data_utils._array_copy_data(res_opt, exhibits_traits, options)
        result.contents = copy_data_utils._array_copy_data(res_opt, instance.contents, options)

        return result

File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/CdmFolder/AttributeContextPersistence.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder.Types;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Newtonsoft.Json.Linq;
    using System.Linq;

    class AttributeContextPersistence
    {
        public static CdmAttributeContext FromData(CdmCorpusContext ctx, dynamic obj)
        {
            if (obj == null)
            {
                return null;
            }

            CdmAttributeContext attributeContext = ctx.Corpus.MakeObject<CdmAttributeContext>(CdmObjectType.AttributeContextDef, obj.Value<string>("name"), false);
            attributeContext.Type = MapTypeNameToEnum(obj.Value<string>("type"));
            if (obj.Value<string>("parent") != null)
                attributeContext.Parent = AttributeContextReferencePersistence.FromData(ctx, obj.Value<string>("parent"));
            string explanation = obj.Value<string>("explanation");
            if (!string.IsNullOrEmpty(explanation))
                attributeContext.Explanation = explanation;
            if (obj.Value<string>("definition") != null)
            {
                switch (attributeContext.Type)
                {
                    case CdmAttributeContextType.Entity:
                    case CdmAttributeContextType.EntityReferenceExtends:
                        attributeContext.Definition = EntityReferencePersistence.FromData(ctx, obj.Value<string>("definition"));
                        break;
                    case CdmAttributeContextType.AttributeGroup:
                        attributeContext.Definition = AttributeGroupReferencePersistence.FromData(ctx, obj.Value<string>("definition"));
                        break;
                    case CdmAttributeContextType.AddedAttributeSupporting:
                    case CdmAttributeContextType.AddedAttributeIdentity:
                    case CdmAttributeContextType.AddedAttributeExpansionTotal:
                    case CdmAttributeContextType.AddedAttributeSelectedType:
                    case CdmAttributeContextType.AttributeDefinition:
                        attributeContext.Definition = AttributeReferencePersistence.FromData(ctx, obj.Value<string>("definition"));
                        break;
                }
            }
            // i know the trait collection names look wrong. but I wanted to use the def baseclass
            Utils.AddListToCdmCollection(attributeContext.ExhibitsTraits, Utils.CreateTraitReferenceList(ctx, obj.Value<JToken>("appliedTraits")));
            if (obj.Value<JToken>("contents") != null)
            {
                for (int i = 0; i < obj.Value<JToken>("contents").Count; i++)
                {
                    JToken ct = obj.Value<JToken>("contents")[i];
                    if (ct is JValue)
                        attributeContext.Contents.Add(AttributeReferencePersistence.FromData(ctx, (string)ct));
                    else
                        attributeContext.Contents.Add(FromData(ctx, ct));
                }
            }
            if (obj.Value<JToken>("lineage") != null)
            {
                attributeContext.Lineage = new CdmCollection<CdmAttributeContextReference>(ctx, attributeContext, CdmObjectType.AttributeContextRef);
                for (int i = 0; i < obj.Value<JToken>("lineage").Count; i++)
                {
                    JToken ct = obj.Value<JToken>("lineage")[i];
                    attributeContext.Lineage.Add(AttributeContextReferencePersistence.FromData(ctx, ct));
                }
            }

            return attributeContext;
        }

        public static dynamic ToData(CdmAttributeContext instance, ResolveOptions resOpt, CopyOptions options)
        {
            return new AttributeContext
            {
                Explanation = instance.Explanation,
                Name = instance.Name,
                Type = MapEnumToTypeName(instance.Type.Value),
                Parent = instance.Parent?.CopyData(resOpt, options) as string,
                Definition = instance.Definition?.CopyData(resOpt, options) as string,
                // i know the trait collection names look wrong. but I wanted to use the def baseclass
                AppliedTraits = Utils.ListCopyData<dynamic>(resOpt, instance.ExhibitsTraits?.Where(trait => !trait.IsFromProperty)?.ToList(), options)?.ConvertAll<JToken>(t => JToken.FromObject(t, JsonSerializationUtil.JsonSerializer)),
                Contents = Utils.ListCopyData<dynamic>(resOpt, instance.Contents, options)?.ConvertAll<JToken>(t => JToken.FromObject(t, JsonSerializationUtil.JsonSerializer)),
                Lineage = Utils.ListCopyData<dynamic>(resOpt, instance.Lineage, options)?.ConvertAll<JToken>(t => JToken.FromObject(t, JsonSerializationUtil.JsonSerializer))
            };
        }

        private static CdmAttributeContextType? MapTypeNameToEnum(string typeName)
        {
            switch (typeName)
            {
                case "entity":
                    return CdmAttributeContextType.Entity;
                case "entityReferenceExtends":
                    return CdmAttributeContextType.EntityReferenceExtends;
                case "attributeGroup":
                    return CdmAttributeContextType.AttributeGroup;
                case "attributeDefinition":
                    return CdmAttributeContextType.AttributeDefinition;
                case "addedAttributeSupporting":
                    return CdmAttributeContextType.AddedAttributeSupporting;
                case "addedAttributeIdentity":
                    return CdmAttributeContextType.AddedAttributeIdentity;
                case "addedAttributeExpansionTotal":
                    return CdmAttributeContextType.AddedAttributeExpansionTotal;
                case "addedAttributeSelectedType":
                    return CdmAttributeContextType.AddedAttributeSelectedType;
                case "generatedRound":
                    return CdmAttributeContextType.GeneratedRound;
                case "generatedSet":
                    return CdmAttributeContextType.GeneratedSet;
                case "projection":
                    return CdmAttributeContextType.Projection;
                case "source":
                    return CdmAttributeContextType.Source;
                case "operations":
                    return CdmAttributeContextType.Operations;
                case "operationAddCountAttribute":
                    return CdmAttributeContextType.OperationAddCountAttribute;
                case "operationAddSupportingAttribute":
                    return CdmAttributeContextType.OperationAddSupportingAttribute;
                case "operationAddTypeAttribute":
                    return CdmAttributeContextType.OperationAddTypeAttribute;
                case "operationExcludeAttributes":
                    return CdmAttributeContextType.OperationExcludeAttributes;
                case "operationArrayExpansion":
                    return CdmAttributeContextType.OperationArrayExpansion;
                case "operationCombineAttributes":
                    return CdmAttributeContextType.OperationCombineAttributes;
                case "operationRenameAttributes":
                    return CdmAttributeContextType.OperationRenameAttributes;
                case "operationReplaceAsForeignKey":
                    return CdmAttributeContextType.OperationReplaceAsForeignKey;
                case "operationIncludeAttributes":
                    return CdmAttributeContextType.OperationIncludeAttributes;
                case "operationAddAttributeGroup":
                    return CdmAttributeContextType.OperationAddAttributeGroup;
                default:
                    return CdmAttributeContextType.Unknown;
            }
        }

        private static string MapEnumToTypeName(CdmAttributeContextType enumVal)
        {
            switch (enumVal)
            {
                case CdmAttributeContextType.Entity:
                    return "entity";
                case CdmAttributeContextType.EntityReferenceExtends:
                    return "entityReferenceExtends";
                case CdmAttributeContextType.AttributeGroup:
                    return "attributeGroup";
                case CdmAttributeContextType.AttributeDefinition:
                    return "attributeDefinition";
                case CdmAttributeContextType.AddedAttributeSupporting:
                    return "addedAttributeSupporting";
                case CdmAttributeContextType.AddedAttributeIdentity:
                    return "addedAttributeIdentity";
                case CdmAttributeContextType.AddedAttributeExpansionTotal:
                    return "addedAttributeExpansionTotal";
                case CdmAttributeContextType.AddedAttributeSelectedType:
                    return "addedAttributeSelectedType";
                case CdmAttributeContextType.GeneratedRound:
                    return "generatedRound";
                case CdmAttributeContextType.GeneratedSet:
                    return "generatedSet";
                case CdmAttributeContextType.Projection:
                    return "projection";
                case CdmAttributeContextType.Source:
                    return "source";
                case CdmAttributeContextType.Operations:
                    return "operations";
                case CdmAttributeContextType.OperationAddCountAttribute:
                    return "operationAddCountAttribute";
                case CdmAttributeContextType.OperationAddSupportingAttribute:
                    return "operationAddSupportingAttribute";
                case CdmAttributeContextType.OperationAddTypeAttribute:
                    return "operationAddTypeAttribute";
                case CdmAttributeContextType.OperationExcludeAttributes:
                    return "operationExcludeAttributes";
                case CdmAttributeContextType.OperationArrayExpansion:
                    return "operationArrayExpansion";
                case CdmAttributeContextType.OperationCombineAttributes:
                    return "operationCombineAttributes";
                case CdmAttributeContextType.OperationRenameAttributes:
                    return "operationRenameAttributes";
                case CdmAttributeContextType.OperationReplaceAsForeignKey:
                    return "operationReplaceAsForeignKey";
                case CdmAttributeContextType.OperationIncludeAttributes:
                    return "operationIncludeAttributes";
                case CdmAttributeContextType.OperationAddAttributeGroup:
                    return "operationAddAttributeGroup";
                default:
                    return "unknown";
            }
        }

    }
}
